internal/state/listener.go

```diff
@@ -47,93 +57,119 @@ func newListenerConfiguratorFactory(
 ) *listenerConfiguratorFactory {
 	return &listenerConfiguratorFactory{
 		https: newHTTPSListenerConfigurator(gw, secretMemoryMgr),
-		http:  newHTTPListenerConfigurator(),
+		http:  newHTTPListenerConfigurator(gw),
 	}
 }
 
-type httpsListenerConfigurator struct {
+type httpListenerConfigurator struct {
 	gateway         *v1beta1.Gateway
 	secretMemoryMgr SecretDiskMemoryManager
 	usedHostnames   map[string]*listener
+	validate        func(gl v1beta1.Listener) []conditions.Condition
+}
+
+func newHTTPListenerConfigurator(gw *v1beta1.Gateway) *httpListenerConfigurator {
+	return &httpListenerConfigurator{
+		usedHostnames: make(map[string]*listener),
+		gateway:       gw,
+		validate:      validateHTTPListener,
+	}
 }
 
 func newHTTPSListenerConfigurator(
 	gateway *v1beta1.Gateway,
 	secretMemoryMgr SecretDiskMemoryManager,
-) *httpsListenerConfigurator {
-	return &httpsListenerConfigurator{
+) *httpListenerConfigurator {
+	return &httpListenerConfigurator{
 		gateway:         gateway,
 		secretMemoryMgr: secretMemoryMgr,
 		usedHostnames:   make(map[string]*listener),
+		validate: func(gl v1beta1.Listener) []conditions.Condition {
+			return validateHTTPSListener(gl, gateway.Namespace)
+		},
 	}
 }
 
-func (c *httpsListenerConfigurator) configure(gl v1beta1.Listener) *listener {
-	var path string
-	var err error
-
-	valid := validateHTTPSListener(gl, c.gateway.Namespace)
+func validateListener(
+	gl v1beta1.Listener,
+	gw *v1beta1.Gateway,
+	validate func(gl v1beta1.Listener) []conditions.Condition,
+) (conds []conditions.Condition, validHostname bool) {
+	conds = validate(gl)
 
-	if valid {
-		nsname := types.NamespacedName{
-			Namespace: c.gateway.Namespace,
-			Name:      string(gl.TLS.CertificateRefs[0].Name),
-		}
+	if len(gw.Spec.Addresses) > 0 {
```

Comment:
If this is true, doesn't it invalidate all the listeners in the Gateway? Why perform this check for every listener? Do we have to keep processing all the listeners in this case?

Reply:
And we have to generate all possible conditions even if we know the entire resource is invalid?

Reply:
it is undefined by the Gateway API spec

Reply:
Interesting. If it's undefined, I wonder if it makes sense to stop processing here and only report the UnsupportedAddress condition.

Reply:
I think the more problems we can report, the better